### ISSUE TITLE
Unequal padding for MaxPooling

### ISSUE THREAD

**Issue by OverLordGoldDragon** [pytorch/pytorch#71362]
### 🚀 The feature, motivation and pitch

`F.pad` can't serve as a workaround, need native support for padding each side differently in max pooling (tuple per dim).

```python
import torch, torch.nn.functional as F

x = torch.tensor([[-1., -2.]])
mp = lambda x, p=0: F.max_pool1d(x, 2, 2, padding=p)
print(mp(x, 1), mp(F.pad(x, (1, 1))))
```
```
tensor([[-1., -2.]]) tensor([[0., 0.]])
```

cc @albanD @mruberry @jbschlosser @walterddr @kshitij12345

**Comment by albanD** [comment]
Hi,

Could you clarify what is not working as expected here?
Also if you want the padded values to be ignored by the maxpooling, you want to pass `value=float("-inf")` to the pad function.

**Comment by OverLordGoldDragon** [comment]
Max pooling can only pad equally on both sides. There is no `(2, 3)` for `MaxPool1d` as there is for `Conv1d`, for example.

**Comment by jbschlosser** [comment]
Thanks for the report! Note that @albanD's suggestion should function as a workaround:
```python
import torch, torch.nn.functional as F

x = torch.tensor([[-1., -2.]])
padded_x = F.pad(x, (1, 2), value=float('-inf'))  # note unequal padding
out = F.max_pool1d(padded_x, 2, 2)
print(out)
```

**Comment by OverLordGoldDragon** [comment]
Not from a performance standpoint: I have several max pools, which add significant overhead with several pads. I assume internally there's no padding at all with `padding=`, just writing directly to output tensor. Otherwise, sure.

**Comment by albanD** [comment]
AFAIK, anything that isn't symmetric zero padding always does pad+op (where op is conv/pooling etc).
So it is actually doing the padding in most cases.

It would be a significant amount of work to have kernels that handle all these special padding. I don't think we have anyone on the core side working on that.

**Comment by OverLordGoldDragon** [comment]
I haven't benched, but if that's the case, fair enough. Unless required for backprop, it could really benefit performance to "pad without padding" where possible.
